hpx/lcos/detail/future_data.hpp

```diff
@@ -838,13 +837,11 @@ namespace detail
         template <typename T>
         void set_data(T && result)
         {
-            HPX_ASSERT(started_);
-            this->future_data<Result>::set_value(std::forward<T>(result));
+            this->future_data<Result>::set_data(std::forward<T>(result));
         }
 
         void set_exception(boost::exception_ptr const& e)
         {
-            HPX_ASSERT(started_);
```

Comment:
The (remote) `lcos::promise` has to support the deferred policy (that's the point of this PR), so it starts the task only when somebody asks for the value. For this reason the task has not started yet.